x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/analysis/Analyzer.java

```diff
@@ -120,4 +124,24 @@ protected LogicalPlan rule(LogicalPlan plan) {
             });
         }
     }
-}
\ No newline at end of file
+
+    private class MatchLiteralsOnTheRight extends AnalyzerRule<LogicalPlan> {
+
+        @Override
+        protected LogicalPlan rule(LogicalPlan plan) {
+            return plan.transformExpressionsUp(e -> {
+                if (e instanceof Match) {
+                    Match m = (Match) e;
+                    int size = m.children().size();
+                    if ((m.children().get(size - 1) instanceof Literal) == false) {
+                        List<Expression> newChildren = new ArrayList<>(m.children().size());
+                        newChildren.add(m.children().get(size - 1));
+                        newChildren.addAll(m.children().subList(0, size - 1));
+                        return m.replaceChildren(newChildren);
+                    }
+                }
+                return e;
+            });
+        }
+    }
+}
```

Comment:
This should be an optimization rule not analysis (since the literal might not be folded).
It also seems incomplete - what if there are multiple non-literals? Is there any validation happening (verifier rule) in the chain?

Reply:
I didn't like that either to take place in the analysis, but this way the Verifier will call the resolveType() on the Match which will indeed check that the patterns are constants (foldable).
If we apply the rule in the Optimizer that's after the verifier so it will "blow up".  Any suggestion?

Reply:
It sounds to me like the `resolveType` method needs improving;
this rule moves the last items to the front of the list so instead of doing that outside `Match`, the class itself should have the proper behavior in place.